source/common/network/lc_trie.h

```diff
@@ -0,0 +1,487 @@
+        } else if (length_ < b.length_) {
+          return -1;
+        } else if (length_ > b.length_) {
+          return 1;
+        } else {
+          return 0;
+        }
+      }
+    }
+
+    bool operator<(const IpPrefixes& b) const { return (this->compare(b) == -1); }
+
+    /**
+     * @return true if other is a prefix of this.
+     */
+    bool isPrefix(const IpPrefixes& other) {
+      return (length_ == 0 || (length_ <= other.length_ &&
+                               extractBits<IpType, address_size>(0, length_, ip_) ==
+                                   extractBits<IpType, address_size>(0, length_, other.ip_)));
+    }
+
+    // The address represented either in uint32_t or uint128.
+    IpType ip_;
+    // Length of the cidr range.
+    int length_;
+    // TODO(ccaraman): Support more than one tag per entry.
+    // Tag for this entry.
+    std::string tag_;
+  };
+
+  template <class IpType, uint32_t address_size = 8 * sizeof(IpType)> class LcTrieInternal {
+  public:
+    /**
+     * Level Compressed Trie (LC-Trie) implementation for IP(IPv4 or IPv6) prefixes.
+     *
+     * The C++ implementation of the LcTrie found at
+     * https://github.com/beevek/libkrb/blob/master/krb/lc_trie.hpp was used as a reference.
+     * The implemenation above is based off of S. Nilson's code at
+     * http://www.csc.kth.se/~snilsson/software/router/C/
+     *
+     * Note: This trie handles Ipv4 and Ipv6 separately.
+     * Note: The trie can only support up 524288(2^19) nodes. Please see LcNode for further
+     * explanation.
+     *
+     * @param fill_factor supplies the fraction of completeness to use when considering
+     *        if a subtrie should be compressed.
+     * @param root_branching_factor supplies the branching factor at the root. This is
+     *        an optimization as suggested in the paper.
+     * @param tag_data supplies the ip tag and the prefixes for associated with that tag.
+     */
+    LcTrieInternal(std::vector<IpPrefixes<IpType>>& tag_data, double fill_factor = 0.5,
+                   uint32_t root_branching_factor = 0);
+
+    /**
+     * Search the LC Trie for a prefix that has the input, ip, in range.
+     * @param  ip supplies the IP address searching for a prefix in string format.
+     * @return tag from the prefix that encompasses the input. An empty string is returned
+     *         if no prefix in the LC Trie exists.
+     */
+    std::string search(const IpType& ip_address) const;
+
+  private:
+    /**
+     * Builds the Level Compresesed Trie, by first sorting the base vector, removing nested prefixes
+     * and duplicated and then recursively builds the trie by invoking buildRecursively().
+     */
+    void buildLcTrie(std::vector<IpPrefixes<IpType>>& tag_data) {
+      if (tag_data.empty()) {
+        return;
+      }
+
+      // The node used for the trie uses the last 20 bits of uint32_t. The trie cannot support
+      // an input greater than 524288(2^19) nodes.
```

Comment:
I think it might be possible for a trie built from 2^19 prefixes to end up with more than 2^20 nodes, if the fill_factor is set to something less than 0.5. Hmmm, maybe the safe limit is 2^20 * fill_factor, but I haven't thought through it fully, so please don't take my word for it. :)